### ISSUE TITLE
Disallow resource methods from returning values belonging to function and client object type

### ISSUE THREAD

**Issue by rdulmina** [ballerina-platform/ballerina-lang#36630]
**Description:**

According to [spec](https://ballerina.io/spec/lang/master/#resource-method-decl) 
> The return type of a resource method must not allow values belonging to the function basic type and must not allow client object types.

**Comment by rdulmina** [comment]
Assigned to @SasinduDilshara

**Comment by KRVPerera** [comment]
@rdulmina

```Ballerina
type C  client object {
   resource function put x();
   resource function get x() returns client object {
          resource function get y() returns string;
   }

resource function get x() returns client object { // A
        resource function put y() returns string;
   }
};

resource function put x/y() returns client object { // B
 };


C c;

string s = c->/x/y;
```


Do we need to check unions specifically? I think for unions we need to do a cast not allowing to use of the same semantics.

**Comment by rdulmina** [comment]
> @rdulmina

> Do we need to check unions specifically? I think for unions we need to do a cast not allowing to use of the same semantics.

That's how I feel also for client object types after reading the spec sentence `The return type of a resource method must not allow values belonging to the function basic type and must not allow client object types`. It says `values belonging to the function basic type` which means we need to consider unions also. But in the latter part it does not talk about the value it says `must not allow client object types` In which we don't need to check unions specifically. @hasithaa WDYT?

**Comment by KRVPerera** [comment]
We should be able to call this?

```Ballerina
type C  client object {
    function x() returns client object {
         function y() returns string;
    } | client object {
         function y() returns string;
    };
 };

public function main() {
   C c;
  string s = c.x().y(); 
}
```

**Comment by rdulmina** [comment]
Fixed with  #36959
